### PR TITLE
#516 some optionals

### PR DESCRIPTION
**Pull request by DeepSnowNeeL**
WIP

@matozoid is this okay to you ? if yes i'll continue in this way for the rest of "// TODO nullable" I can find

### CODE REVIEW

File: javaparser-core/src/main/java/com/github/javaparser/ast/visitor/CloneVisitor.java [hunk 1]
```diff
@@ -162,7 +162,7 @@ public Visitable visit(AnnotationDeclaration _n, Object _arg) {
 	public Visitable visit(AnnotationMemberDeclaration _n, Object _arg) {
 		NodeList<AnnotationExpr> annotations = cloneList(_n.getAnnotations(), _arg);
 		Type<?> type_ = cloneNode(_n.getType(), _arg);
-		Expression defaultValue = cloneNode(_n.getDefaultValue(), _arg);
+		Expression defaultValue = cloneNode(_n.getDefaultValue().orElse(null), _arg);
```
Comment (matozoid) [inline javaparser-core/src/main/java/com/github/javaparser/ast/visitor/CloneVisitor.java]: Consider making an optional-specific "cloneNode" for this

**Comment by matozoid** [comment]
Yep, that looks like it :) Good luck, I know how much fun it is to update those visitors every time.

File: javaparser-core/src/main/java/com/github/javaparser/ast/nodeTypes/NodeWithStatements.java [hunk 2]
```diff
@@ -71,8 +70,20 @@ default ExpressionStmt addAndGetStatement(Expression expr) {
...
-    
+
     default boolean isEmpty() {
         return getStmts().isEmpty();
     }
+
+    @SuppressWarnings("unchecked")
+    default N copyStatements(NodeList<Statement> nodeList) {
+        for (Statement n : nodeList) {
+            addStatement((Statement) n.clone());
```
Comment (DeepSnowNeeL) [inline javaparser-core/src/main/java/com/github/javaparser/ast/nodeTypes/NodeWithStatements.java:81]: It actually make a shallow copy, do we have something to deep copy statements ? (I tried using CloneVisitor but without success)

Comment (matozoid) [reply javaparser-core/src/main/java/com/github/javaparser/ast/nodeTypes/NodeWithStatements.java:81]: Why was clonevisitor not a success? It's in use behind n.clone()

Comment (DeepSnowNeeL) [reply javaparser-core/src/main/java/com/github/javaparser/ast/nodeTypes/NodeWithStatements.java:81]: Oh yeah ! I thought it was using java standard clone() for a moment ! so it is a deep copy, great

**Comment by matozoid** [comment]
Are you also going to do the Javadoc? ;)

Actually, it would be good to at least document the setters that can take a null, with a little explanation of what null means in that case. "Null means no initializer" or so. In other cases it would be "Null is not allowed"

**Comment by DeepSnowNeeL** [comment]
Well I did the javadoc for the setter, but yep i'll try to be precise in
what null means

Le dim. 6 nov. 2016 à 16:37, Danny van Bruggen notifications@github.com a
écrit :

**Comment by DeepSnowNeeL** [comment]
I'm almost done but @matozoid can you check the 1 last test, I can't get what's the difference, maybe a CR LF windows problem ?

**Comment by matozoid** [comment]
Okay, I'll see.

**Comment by matozoid** [comment]
It's working fine for me. Don't look at appveyor if that's what you're worried about. It hasn't been working for months now :(

**Comment by matozoid** [comment]
I see you also did a lot of the reformatting job :D

Double check the Javadoc for NodeWithTypeArguments please.

**Comment by DeepSnowNeeL** [comment]
Well my eclipse always reformat / re-import and there is no way to disable that unless I ctrl+Z after each save (it undo the organize import)

What about the NodeWithTypeArguments ?

**Comment by matozoid** [comment]
NodeWithTypeArguments: it mentions null in the Javadoc.

I used to use Eclipse, and you can turn that off :) But never mind, you improved the formatting.

**Comment by DeepSnowNeeL** [comment]
done :)

**Comment by matozoid** [comment]
Merge time?

**Comment by matozoid** [comment]
Thank you very much for your contribution :)

**Pull request merged**